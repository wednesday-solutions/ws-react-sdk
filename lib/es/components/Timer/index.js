import{_ as e,d as n}from"../_rollupPluginBabelHelpers-c2a46d82.js";import"../_rollup-plugin-inject-process-env-0dfb24f4.js";import r,{useState as t,useEffect as i}from"react";import a from"prop-types";import l from"styled-components";import o from"../T/index.js";import{c as s}from"../colors-bf7c26a7.js";import{f as c}from"../fonts-56bad7cc.js";import"react-intl";import"../If/index.js";import"../media-54d0e07e.js";var p,m,d=l(o)(p||(p=e(["\n  color: ",";\n  ",";\n  ",";\n  line-height: 1.5;\n  letter-spacing: 1.2px;\n  margin: 0;\n  opacity: ",";\n  &:hover {\n    cursor: ",";\n  }\n"])),s.primary,c.weights.bold(),c.size.xSmall(),(e=>e.disable?.5:1),(e=>e.disable?"not-allowed":"pointer")),u=l.p(m||(m=e(["\n  color: ",";\n  ",";\n  ",";\n  line-height: 1.5;\n  letter-spacing: 1.2px;\n  opacity: 0.5;\n  margin: 0;\n"])),s.secondary,c.weights.bold(),c.size.xSmall());function f(e){var{initialMinute:a,initialSeconds:l,retryNo:o,resendCallback:s}=e,[c,p]=t(a),[m,f]=t(l),[b,y]=t(o);function g(){return(g=n((function*(){if(b>0)try{yield s(),y((e=>e-1)),p(a),f(l)}catch(e){}}))).apply(this,arguments)}return i((()=>{var e=setInterval((()=>{m>0&&f(m-1),0===m&&(0===c?clearInterval(e):(p(c-1),f(59)))}),1e3);return()=>{clearInterval(e)}})),r.createElement("div",{"data-testid":"timer"},0===c&&0===m?r.createElement(d,{id:"resend",type:"subText",onClick:function(){return g.apply(this,arguments)},disable:b<=0}):r.createElement(u,null," ",c,":",m<10?"0".concat(m):m))}f.propTypes={initialMinute:a.number,initialSeconds:a.number,retryNo:a.number,resendCallback:a.func},f.defaultProps={initialMinute:0,initialSeconds:0,retryNo:0,resendCallback:()=>{}};var b=f;export default b;
