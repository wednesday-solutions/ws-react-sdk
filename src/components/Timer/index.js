/**
 *
 * Timer
 *
 */

import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import T from '@components/T';
import { colors, fonts } from '@themes';

const Resend = styled(T)`
  color: ${colors.primary};
  ${fonts.weights.bold()};
  ${fonts.size.xSmall()};
  line-height: 1.5;
  letter-spacing: 1.2px;
  margin: 0;
  opacity: ${props => (props.disable ? 0.5 : 1)};
  &:hover {
    cursor: ${props => (props.disable ? 'not-allowed' : 'pointer')};
  }
`;
const Clock = styled.p`
  color: ${colors.secondary};
  ${fonts.weights.bold()};
  ${fonts.size.xSmall()};
  line-height: 1.5;
  letter-spacing: 1.2px;
  opacity: 0.5;
  margin: 0;
`;

function Timer({ initialMinute, initialSeconds, retryNo, resendCallback }) {
  const [minutes, setMinutes] = useState(initialMinute);
  const [seconds, setSeconds] = useState(initialSeconds);
  const [retry, setRetry] = useState(retryNo);

  async function resendHandler() {
    if (retry > 0) {
      try {
        await resendCallback();
        setRetry(n => n - 1);
        setMinutes(initialMinute);
        setSeconds(initialSeconds);
      } catch (e) {}
    }
  }

  useEffect(() => {
    const timerInterval = setInterval(() => {
      if (seconds > 0) {
        setSeconds(seconds - 1);
      }
      if (seconds === 0) {
        if (minutes === 0) {
          clearInterval(timerInterval);
        } else {
          setMinutes(minutes - 1);
          setSeconds(59);
        }
      }
    }, 1000);
    return () => {
      clearInterval(timerInterval);
    };
  });

  return (
    <div data-testid="timer">
      {minutes === 0 && seconds === 0 ? (
        <Resend id="resend" type="subText" onClick={resendHandler} disable={retry <= 0} />
      ) : (
        <Clock>
          {' '}
          {minutes}:{seconds < 10 ? `0${seconds}` : seconds}
        </Clock>
      )}
    </div>
  );
}

Timer.propTypes = {
  initialMinute: PropTypes.number,
  initialSeconds: PropTypes.number,
  retryNo: PropTypes.number,
  resendCallback: PropTypes.func
};

Timer.defaultProps = {
  initialMinute: 0,
  initialSeconds: 0,
  retryNo: 0,
  resendCallback: () => {}
};

const TimerComponent = Timer;
export default TimerComponent;
