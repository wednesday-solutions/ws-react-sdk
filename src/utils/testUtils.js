import React from 'react';
import { IntlProvider } from 'react-intl';
import { render } from '@testing-library/react';
import { DEFAULT_LOCALE, translationMessages } from '@src/i18n';
import IntlGlobalProvider from '@components/IntlGlobalProvider';

export const renderWithIntl = children => {
  return render(
    <IntlProvider locale={DEFAULT_LOCALE} messages={translationMessages[DEFAULT_LOCALE]}>
      <IntlGlobalProvider>{children}</IntlGlobalProvider>
    </IntlProvider>
  );
};

export const getComponentStyles = (Component, props = {}) => {
  renderWithIntl(Component(props));
  const { styledComponentId } = Component(props).type;
  const componentRoots = document.getElementsByClassName(styledComponentId);
  // eslint-disable-next-line no-underscore-dangle
  return window.getComputedStyle(componentRoots[0])._values;
};

export const timeout = ms => new Promise(resolve => setTimeout(resolve, ms));
export const apiResponseGenerator = (ok, data) => ({
  ok,
  data
});

export default {
  timeout,
  renderWithIntl,
  getComponentStyles,
  apiResponseGenerator
};
