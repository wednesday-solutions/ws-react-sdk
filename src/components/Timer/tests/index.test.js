/**
 *
 * Tests for Timer
 *
 */

import React from 'react';
import { fireEvent } from '@testing-library/dom';
import { renderWithIntl, timeout } from '@utils/testUtils';
import Timer from '../index';

describe('<Timer />', () => {
  it('should render and match the snapshot', () => {
    const { baseElement } = renderWithIntl(<Timer />);
    expect(baseElement).toMatchSnapshot();
  });

  it('should contain 1 Timer component', () => {
    const { getAllByTestId } = renderWithIntl(<Timer />);
    expect(getAllByTestId('timer').length).toBe(1);
  });

  it('should render resend in 1 min', async () => {
    const { getAllByText } = renderWithIntl(<Timer initialSeconds={3} retryNo={2} />);
    await timeout(3500);
    expect(getAllByText(/RESEND/).length).toBe(1);
  });
  it('should handle resend button click and match the snapshot', async () => {
    const jestSpy = jest.fn();
    const { getByTestId } = renderWithIntl(
      <Timer initialSeconds={0} initialMinute={0} retryNo={2} resendCallback={jestSpy} />
    );
    fireEvent.click(getByTestId('t'));
    await timeout(3500);
    expect(getByTestId).toMatchSnapshot();
    expect(jestSpy).toBeCalled();
  });

  it('should handle minutes interval and match the snapshot ', async () => {
    const jestSpy = jest.fn();
    const { getByTestId } = renderWithIntl(
      <Timer initialSeconds={0} initialMinute={2} retryNo={2} resendCallback={jestSpy} />
    );
    expect(getByTestId).toMatchSnapshot();
  });
});
