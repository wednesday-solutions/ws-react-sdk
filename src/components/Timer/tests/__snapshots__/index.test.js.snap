// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Timer /> should render and match the snapshot 1`] = `
<body>
  <div>
    <div
      data-testid="timer"
    >
      <p
        class="sc-bdfBwQ gZxDMi sc-gsTCUz bruhyv"
        data-testid="t"
      >
        RESEND
      </p>
    </div>
  </div>
</body>
`;
